src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -217,53 +222,68 @@ public FormPipeReader(PipeReader pipeReader, Encoding encoding)
             bool isFinalBlock)
         {
             var sequenceReader = new SequenceReader<byte>(buffer);
+            ReadOnlySequence<byte> segment;
+
             var consumed = sequenceReader.Position;
             var consumedBytes = default(long);
             var equalsDelimiter = GetEqualsForEncoding();
             var andDelimiter = GetAndForEncoding();
 
             while (!sequenceReader.End)
             {
-                // TODO seems there is a bug with TryReadTo (advancePastDelimiter: true). It isn't advancing past the delimiter on second read.
-                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> key, equalsDelimiter, advancePastDelimiter: false) ||
-                    !sequenceReader.IsNext(equalsDelimiter, true))
+                if (!sequenceReader.TryReadTo(out segment, andDelimiter))
                 {
-                    if ((sequenceReader.Consumed - consumedBytes) > KeyLengthLimit)
+                    if (!isFinalBlock)
                     {
-                        ThrowKeyTooLargeException();
+                        // Don't buffer indefinately
+                        if ((sequenceReader.Consumed - consumedBytes) > KeyLengthLimit + ValueLengthLimit)
+                        {
+                            ThrowKeyOrValueTooLargeException();
+                        }
+                        break;
                     }
 
-                    break;
+                    // This must be the final key=value pair
+                    segment = buffer.Slice(sequenceReader.Position);
+                    sequenceReader.Advance(segment.Length);
                 }
 
-                if (key.Length > KeyLengthLimit)
+                if (segment.IsSingleSegment)
                 {
-                    ThrowKeyTooLargeException();
+                    ParseFormValuesFast(segment.First.Span, ref accumulator, isFinalBlock: true, out var segmentConsumed);
+                    Debug.Assert(segmentConsumed == segment.First.Span.Length);
+                    continue;
                 }
 
-                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> value, andDelimiter, false) ||
-                    !sequenceReader.IsNext(andDelimiter, true))
+                var segmentReader = new SequenceReader<byte>(segment);
```

Comment:
This looks expensive, one sequence reader per loop?

Reply:
I can't rewind the main reader can I? 

It's a ref struct, how expensive are those?

Reply:
You just don't need to call Advance. Save the position of the & and then find the =. Advance afterwards.

Reply:
Let's profile it though, my guess is saving the allocation for the sequence reader is worth it in the long run.

Reply:
Since it's a struct, you can just use assignment to create a copy of it on the stack and create a checkpoint of sorts. I think @jkotalik's suggestion of just not Advancing forward until you find the & and the = might be faster though.

Reply:
doesn't TryReadTo Advance by default? I'm only calling Advance in the negative case.

Reply:
Just looked at it. Sure seems like it does, and I don't see any easy way to tell it not to. Copying a SequenceReader at the beginning saves you a ctor call, but I doubt that has a huge perf impact relative to avoiding the extra struct altogether. I also don't think that helps you  limit your search for = to be before the next &.

Reply:
I think you are correct. There is a rewind method: https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/SequenceReader.cs#L156

Reply:
Interesting. Might be worth benchmarking this vs the rewind solution. I could see the rewind solution being faster if the RetreatToPreviousSpan path is uncommon which I think it would be.

Reply:
Actually wouldn't we always retreat to the previous span as this case isn't within a single span?

Reply:
> Actually wouldn't we always retreat to the previous span as this case isn't within a single span?

Depends on where we find the "&"

Reply:
Maybe we shouldn't do rewind, it looks like it could make this algorithm O(n^2) in the pathological case since when you cross segments to rewind, it has to [rewalk the ReadOnlySequence](https://github.com/dotnet/corefx/blob/739f4437298d3d010f3427f557bf6d189aac67a7/src/System.Memory/src/System/Buffers/SequenceReader.cs#L173-L182) from the very beginning.

Reply:
Yeah, that isn't great. I wish there was a "TryPeekTo" method on SequenceReader s.t. we don't advance the reader. But agree with @halter73 here.